packages/flutter/lib/src/widgets/pop_scope.dart

```diff
@@ -0,0 +1,114 @@
+
+import 'framework.dart';
+import 'navigator.dart';
+import 'routes.dart';
+
+/// A callback type for informing that a navigation pop has happened.
+///
+/// Accepts a success boolean indicating whether or not back navigation
+/// succeeded.
+typedef OnPoppedCallback = void Function(bool success);
+
+/// Manages system back gestures.
+///
+/// [popEnabled] can be used to disable system back gestures, while [onPopped]
+/// reports when they occur.
+///
+/// {@tool dartpad}
+/// This sample demonstrates how to use this widget to properly handle system
+/// back gestures when using nested [Navigator]s.
+///
+/// ** See code in examples/api/lib/widgets/pop_scope/pop_scope.0.dart **
+/// {@end-tool}
+///
+/// See also:
+///
+///  * [NavigatorPopHandler], which is a less verbose way to handle system back
+///    gestures in the case of nested [Navigator]s.
+///  * [Form.popEnabled] and [Form.onPopped], which can be used to handle system
+///    back gestures in the case of a form with unsaved data.
+///  * [ModalRoute.registerPopScope] and [ModalRoute.unregisterPopScope],
+///    which this widget uses to integrate with Flutter's navigation system.
+class PopScope extends StatefulWidget {
+  /// Creates a widget that registers a callback to veto attempts by the user to
+  /// dismiss the enclosing [ModalRoute].
+  ///
+  /// The [child] argument must not be null.
+  const PopScope({
+    super.key,
+    required this.child,
+    this.popEnabled = true,
+    this.onPopped,
+  });
+
+  /// The widget below this widget in the tree.
+  ///
+  /// {@macro flutter.widgets.ProxyWidget.child}
+  final Widget child;
+
+  /// Called after a route pop was handled.
+  ///
+  /// It's not possible to prevent the pop from happening at the time that this
+  /// method is called; the pop has already happened. Use [popEnabled] to
+  /// disable pops in advance.
+  ///
+  /// Even when the pop is canceled, such as when [popEnabled] is false, this
+  /// will still be called. The `success` parameter indicates whether or not the
+  /// back navigation actually happened successfully.
+  ///
+  /// See also:
+  ///
+  ///  * [Route.onPopped], which is similar.
+  final OnPoppedCallback? onPopped;
+
+  /// {@template flutter.widgets.PopScope.popEnabled}
+  /// When false, blocks the current route from being popped.
+  ///
+  /// This includes the root route, where upon popping, the Flutter app would
+  /// exit.
+  ///
+  /// If multiple [PopScope] widgets appear in a route's widget subtree, then
+  /// each and every `popEnabled` must be `true` in order for the route to be
+  /// able to pop.
+  ///
+  /// This may have implications for route transitions that allow some
+  /// interaction before committing to popping the route. For example,
+  /// [Android's predictive back](https://developer.android.com/guide/navigation/predictive-back-gesture)
+  /// feature will not animate at all when this boolean is true.
+  /// {@endtemplate}
+  final bool popEnabled;
```

Comment:
should this be called canPop? Since that is used else where as well, and the Route.popEnabled returns RoutePopDisposition instead of bool.

Reply:
Good point...  What will be the difference between canPop and popEnabled then?  Maybe we can sort that out too.

In ModalRoute there was already an existing `canPop` method:

```dart
  bool get canPop => hasActiveRouteBelow || willHandlePopInternally;
```

And then `willPop` had slightly different logic:

```dart
  Future<RoutePopDisposition> willPop() async {
    if (willHandlePopInternally) {
      return RoutePopDisposition.pop;
    }
    return super.willPop();
  }
```

`popEnabled` is pretty much just a reskin of that:

```dart
  RoutePopDisposition popEnabled() {
    if (willHandlePopInternally) {
      return RoutePopDisposition.pop;
    }
    return super.popEnabled();
  }
```

Should popEnabled be using canPop internally?  Or maybe this popEnabled should be named popBehavior or something like that?

Reply:
I agree popEnabled is not a good name given its return type, If I don't care about breaking change, I would change it. same for willPop too.

I still prefer canPop as the name. Also maybe `willHandlePopInternally` should take into account of PopScope?

Another thing is `impliesAppBarDismissal`, this controls whether the app bar back button will show up. I am not sure whether we should provide an API for PopScope to play a part in this, I am ok with leave it out for now until someone ask. But just want to make sure we are aware of it.

Reply:
Here's what we have now after my most recent changes:

  * PopScope.canPop (new, returns bool)
  * Route.canPop (old/unchanged, returns bool)
  * Route.willPop (deprecated, returns RoutePopDisposition)
  * Route.popDisposition (new, returns RoutePopDisposition)